Feature: Variable interface scaling

PeterN opened the pull request:
## Motivation / Problem

Powers-of-2 interface scaling, allowing 1x, 2x, and 4x zoom, is too limited, and does not provide a comfortable middle ground.

Additionally window scaling is currently broken. Although padding is scaled at the nested widget level, individually it is left up to each window to scale appropriately.

## Description

This massive set of changes:

1) Changes many rendering routines to work with `Rect` rather than individual left, top, right and bottom, as supplied to `DrawWidget`.
2) Adds `RectPadding`, and new functions to `Rect` apply padding and Indent/Width operations that understand RTL layout decisions.
3) Replaces the `WidgetDrawDistances` enum with pre-scaled `WidgetDimensions`. As padding is now pre-scaled, window scaling is now far more consistent.

**4) Replaces separate GUI and font zoom choices with a single interface scale slider that ranges from 100% to 500%. This can be set in 1% increments, but the UI is limited to 25% increments for ease of use.**

5) Bevels are now scaled. Chunky. The option to toggle this is not currently saved, and may be removed.

![image](https://user-images.githubusercontent.com/639850/198917760-5a9a59aa-058a-4ac2-92fe-f84142b523e4.png)

## Limitations

It's big. This is actually split into a series of branches in my tree, but PRs don't support that AFAIK.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Feature: Variable GUI scale.

GUI scale is now variable from 100% to 500%, and no longer restricted to
powers-of-2.

PeterN committed:
Codechange: Prefer suggested widget padding.

PeterN committed:
Codechange: Add some Rect helpers.

Shrink/Expand/Translate are marked nodiscard as they provide a new Rect
rather than modifying the original Rect.

PeterN committed:
Codechange: Use Rect in DropDownListItem::Draw().

PeterN committed:
Change: Hide WidgetDrawDistances from general use.

PeterN committed:
Cleanup: Simplify CenterBounds()

PeterN committed:
Change: Use power-of-2 scaling for some dimensions in GUI.

These are related to drawing sprites that are normally drawn on the
landscape, and should therefore still follow power-of-2 scaling to fit
correctly.

PeterN committed:
Change: Use CenterBounds function in more places.

CenterBounds may have a 1 pixel difference so the result is not exactly
the same.

PeterN committed:
Change: Set up suggested widget padding for panel, frame and inset.

Window handlers are free to ignore the suggested padding, however some
handlers have logic that assumes this information was already provided.

PeterN committed:
Codechange: Rename slider widget functions to be less specific.

PeterN committed:
Add: Setting to toggle thin vs chunky (scaled) bevels.

PeterN committed:
Change: Use scaled WidgetDimensions.

PeterN committed:
Change: Ignore offsets when drawing GUI sprite.

PeterN committed:
Add: NWidgetPart SetPadding via RectPadding

PeterN committed:
Fix: Settings GUI sprite size.

PeterN committed:
Change: Use standard dimensions instead of custom widths.

PeterN committed:
Codechange: Helpers to allow passing a Rect to some functions.

PeterN committed:
Add: Define scaled WidgetDimensions.

Includes RectPadding container.

PeterN committed:
Add: RectPadding type.

This is similar to Rect but specifies padding to apply to a Rect.

PeterN committed:
Change: Use RectPadding for widget padding/uz_padding.

PeterN committed:
Change: Use RectPadding Horizontal()/Vertical() helpers.

PeterN committed:
Codechange: Make use of Rect Width/Height helpers.

PeterN committed:
Change: Scale position of font shadow.

PeterN committed:
Change: Split vscroll and hscroll padding.

Vertical and horizontal scrollbars are different sizes, this allows
different padding for each type.

PeterN committed:
Fix: Off-by-one in GetAlignedPosition().

PeterN committed:
Change: Simplify drawing widget image buttons.

PeterN committed:
Change: Use scaled WidgetDimensions for core widgets.

PeterN committed:
Change: Sprite-scale inter-character spacing of fonts.

PeterN committed:
Change: Rearrange layouts of station/depot picker windows.

PeterN committed:
Change: Use Rect helpers for widget drawing.

This replaces repetitive and sometimes unwieldy use of constants.

PeterN committed:
Fix: Engine preview text overly padded.

WIP_EP_QUESTION is already padded at the widget level, so no need to apply
extra padding to left/right.

PeterN committed:
Change: Move error window padding to widget level.

This simplifies calculation during drawing.

PeterN committed:
Change: Use standard 'frametext' padding for settings window.

PeterN committed:
Change: Use Rect for QueryString editor.

PeterN committed:
Change: Align picker previews to button centre.

PeterN committed:
Change: Use standard padding for AI Debug window.

PeterN committed:
Change: Use Rect for drawing vehicle details.

PeterN committed:
Change: Use Rect when drawing infrastructure window.

PeterN committed:
Change: Use Rect when drawing build engine list.

PeterN committed:
Codechange: Add parameters to change range of slider widget.

PeterN committed:
Change: Scale thickness of slider bevel.

PeterN committed:
Change: Put ends of slider under centre of control knob.

PeterN committed:
Change: Add optional text labels to slider widget.

LordAro commented on src/core/geometry_type.hpp:
```diff
@@ -43,12 +43,178 @@ struct Dimension {
+	 * @param s number of pixels to remove from each side of Rect.
+	 * @return the new smaller Rect.
+	 */
+	[[nodiscard]] inline Rect Shrink(int s) const {
```
Opening function braces should be on a newline

LordAro commented on src/settings_gui.cpp:
```diff
@@ -474,10 +466,44 @@ struct GameOptionsWindow : Window {
+				if (click_count > 0) this->mouse_capture_widget = widget;
+				break;
+
+			case WID_GO_GUI_SCALE_AUTO:
```
Needs a FALLTHROUGH here, I think?

rubidium42 commented on src/gfx_func.h:
```diff
@@ -139,7 +139,7 @@ bool FillDrawPixelInfo(DrawPixelInfo *n, int left, int top, int width, int heigh
 static inline int CenterBounds(int min, int max, int size)
 {
-	return min + (max - min - size + 1) / 2;
+	return (min + max - size + 1) / 2;
```
Technically this is less overflow safe; imagine `min` to be 2^30, max 2^30 + 100 and size 50. The old code would return 2^30 + 25, the new code would return -2^30 + 26. However, the old code is not guaranteed to always return the right value, see e.g. https://www.youtube.com/watch?v=sBtAGxBh-XI for the mess you can get with getting the midpoint of two values.
An even simpler version would be `std::midpoint(min, max - size + 1)` (with C++20).
Having said that, I guess it's outrageous to have such large coordinates to draw with, so I guess in the end it's not a big issue.

PeterN replied:
Yeah, I think at the point we are hitting those possibilities it's well outside the range of being displayable. Though this commit itself is not particularly relevant to the rest.

rubidium42 commented on src/core/geometry_type.hpp:
```diff
@@ -49,6 +49,121 @@ struct Rect {
+	 * @param width height in pixels for new Rect.
+	 * @return the new resized Rect.
+	 */
+	inline Rect SetHeight(int height) const
```
Can't you just do `SetHeight(int height, bool end = false)` in the next function and omit this?

rubidium42 commented on src/core/geometry_type.hpp:
```diff
@@ -49,6 +49,121 @@ struct Rect {
+	 * @param end   if set, set width at end of Rect, i.e. on right.
+	 * @return the new resized Rect.
+	 */
+	inline Rect SetWidth(int width, bool end) const
```
These setter names are quite misleading. Maybe use something like `WithWidth`, so you would get `rect.WithWidth(10, false)`?

PeterN replied:
WithWidth/Height works for me, also added [[nodiscard]] hint just in case.

rubidium42 commented on src/core/geometry_type.hpp:
```diff
@@ -49,6 +49,121 @@ struct Rect {
+	 * @param end   if set, set height at end of Rect, i.e. at bottom.
+	 * @return the new resized Rect.
+	 */
+	inline Rect SetHeight(int height, bool end) const
```
`WithHeight` to not mislead people?

rubidium42 commented on src/widgets/slider.cpp:
```diff
@@ -23,12 +23,16 @@ static const int SLIDER_WIDTH = 3;
  * @param value Value to put the slider at
  */
-void DrawSliderWidget(Rect r, int min_value, int max_value, int value)
+void DrawSliderWidget(Rect r, int min_value, int max_value, int value, std::map<int, StringID> labels)
```
`const std::map<int, StringID> &labels` to prevent copying the map for each draw?

rubidium42 commented on src/widgets/slider.cpp:
```diff
@@ -42,10 +46,23 @@ void DrawSliderWidget(Rect r, int min_value, int max_value, int value)
 
+	int x;
+	for (auto label : labels)
+	{
```
The `{` should be on the line before.

PeterN committed:
Change: Standardize progress bar layout.

Progress bars are drawn differently depending on when it was added, with
different layouts and sizes.

This change adds a standard padding size to use, and makes all progress
bars visually similar, with scaled padding.

PeterN committed:
Change: Use modalpopup padding for error/query windows.

LordAro commented on src/company_gui.cpp:
```diff
@@ -159,15 +155,14 @@ static uint GetMaxCategoriesWidth()
 	for (uint i = 0; i < list.length; i++) {
 		et = list.et[i];
-		DrawString(r.left + offs_left, r.right - offs_right, y, STR_FINANCES_SECTION_CONSTRUCTION + et);
+		DrawString(tr.left, tr.right, y, STR_FINANCES_SECTION_CONSTRUCTION + et);
```
Possible future work to change DrawString to take a Rect as well, rather than the left/right/top param?

PeterN replied:
It does, but in this case y/start_y is provided separately. Although `tr.top = start_y;` would work, and then just increment tr.top instead of y.

LordAro commented on src/company_gui.cpp:
```diff
@@ -1269,81 +1262,81 @@ static const NWidgetPart _nested_select_company_manager_face_widgets[] = {
 						EndContainer(),
 						NWidget(NWID_HORIZONTAL),
-							NWidget(WWT_TEXT, INVALID_COLOUR, WID_SCMF_CHIN_TEXT), SetFill(1, 0), SetPadding(0, WD_FRAMERECT_RIGHT, 0, WD_FRAMERECT_LEFT),
+							NWidget(WWT_TEXT, INVALID_COLOUR, WID_SCMF_CHIN_TEXT), SetFill(1, 0), SetPadding(WidgetDimensions::unscaled.framerect),
```
How bad/hard would it be to introduce some defaults, like `SetPadding(WidgetDimensions::unscaled.framerect),` ?

PeterN replied:
Most widgets already have built in padding, but WWT_TEXT doesn't. But in most cases, WWT_TEXT is used without extra padding anyway.

LordAro commented on src/dock_gui.cpp:
```diff
@@ -531,25 +531,34 @@ struct BuildDocksDepotWindow : public PickerWindowBase {
+		switch (widget) {
+			case WID_BDD_X:
+			case WID_BDD_Y:
+				Axis axis = widget == WID_BDD_X ? AXIS_X : AXIS_Y;
```
Variable inside case statement - should be surrounded by braces (it doesn't *strictly* matter because it's the last case statement, but is mildly more correct)

LordAro commented on src/industry_gui.cpp:
```diff
@@ -979,11 +962,11 @@ class IndustryViewWindow : public Window
 		}
 
-		return y + WD_FRAMERECT_BOTTOM;
+		return ir.top - 1 + WidgetDimensions::scaled.framerect.bottom;
```
why -1 ?

PeterN replied:
Because otherwise it would be 1 pixel too large ;-)

(It returns the required 'bottom' position, rather than height)

LordAro commented on src/lang/english.txt:
```diff
@@ -1008,21 +1008,19 @@ STR_GAME_OPTIONS_VIDEO_VSYNC_TOOLTIP                            :{BLACK}Check th
+STR_GAME_OPTIONS_GUI_SCALE_2X                                   :{TINY_FONT}2x
+STR_GAME_OPTIONS_GUI_SCALE_3X                                   :{TINY_FONT}3x
+STR_GAME_OPTIONS_GUI_SCALE_4X                                   :{TINY_FONT}4x
+STR_GAME_OPTIONS_GUI_SCALE_5X                                   :{TINY_FONT}5x
```
Why put TINY_FONT in the language file? We're not expecting translations to use something else, surely?

PeterN replied:
No but we want the string to be tiny. I suppose I was thinking a bit to generically with the list of strings, but given the use-case I guess I could just specify FS_SMALL when drawing.

PeterN replied:
Well, it turns out `GetStringBoundingBox` with a StringID doesn't accept a font-size parameter, only with a raw string.

LordAro commented on src/misc_gui.cpp:
```diff
@@ -798,39 +796,36 @@ void QueryString::DrawEditBox(const Window *w, int wid) const
 	 * space reserved at the end for the caret to show */
 	const Textbuf *tb = &this->text;
-	int delta = std::min(0, (right - left) - tb->pixels - 10);
+	int delta = std::min(0, (fr.right - fr.left) - tb->pixels - 10);
```
fr.Width() ?

PeterN replied:
Not sure, Width() adds 1, it's not the same as `right - left`. Maybe the existing code is obiwan though.

LordAro commented on src/misc_gui.cpp:
```diff
@@ -856,17 +851,16 @@ Point QueryString::GetCaretPosition(const Window *w, int wid) const
 	/* Clamp caret position to be inside out current width. */
 	const Textbuf *tb = &this->text;
-	int delta = std::min(0, (right - left) - tb->pixels - 10);
+	int delta = std::min(0, (r.right - r.left) - tb->pixels - 10);
```
r.Width() ?

LordAro commented on src/network/network_chat_gui.cpp:
```diff
@@ -106,7 +107,7 @@ void CDECL NetworkAddChatMessage(TextColour colour, uint duration, const std::st
 {
 	_chatmsg_box.y       = 3 * FONT_HEIGHT_NORMAL;
-	_chatmsg_box.height  = MAX_CHAT_MESSAGES * (FONT_HEIGHT_NORMAL + NETWORK_CHAT_LINE_SPACING) + 4;
+	_chatmsg_box.height  = MAX_CHAT_MESSAGES * (FONT_HEIGHT_NORMAL + ScaleGUITrad(NETWORK_CHAT_LINE_SPACING)) + ScaleGUITrad(4);
```
4, 10 & 3 in this file could be replaced with constants?

LordAro commented on src/newgrf_gui.cpp:
```diff
@@ -833,21 +834,22 @@ struct NewGRFWindow : public Window, NewGRFScanCallback {
-				uint text_right   = rtl ? r.right - square.width - 15 : r.right - WD_FRAMERECT_RIGHT;
-				uint square_left  = rtl ? r.right - square.width - 5 : r.left + 5;
-				uint warning_left = rtl ? r.right - square.width - warning.width - 10 : r.left + square.width + 10;
+				uint text_left    = rtl ? tr.left : tr.left + square.width + 13;
+				uint text_right   = rtl ? tr.right - square.width - 13 : tr.right;
+				uint square_left  = rtl ? tr.right - square.width - 3 : tr.left + 3;
+				uint warning_left = rtl ? tr.right - square.width - warning.width - 8 : tr.left + square.width + 8;
```
scary arbitrary numbers

LordAro commented on src/rail_gui.cpp:
```diff
@@ -1254,11 +1254,11 @@ struct BuildRailStationWindow : public PickerWindowBase {
 					_cur_dpi = &tmp_dpi;
-					int x = ScaleGUITrad(31) + 1;
-					int y = r.bottom - r.top - ScaleGUITrad(31);
+					int x = (r.Width()  - ScaleSpriteTrad(64)) / 2 + ScaleSpriteTrad(31);
+					int y = (r.Height() + ScaleSpriteTrad(58)) / 2 - ScaleSpriteTrad(31);
```
64/58 scalings used in a few places. Pull it out into a variable?

LordAro commented on src/rail_gui.cpp:
```diff
@@ -1673,15 +1673,14 @@ struct BuildSignalWindow : public PickerWindowBase {
-				(widget->current_y - WD_IMGBTN_TOP - WD_IMGBTN_BOTTOM + sig_sprite_size.height) / 2; // aligned to bottom
+		Rect ir = r.Shrink(WidgetDimensions::scaled.imgbtn);
+		int x = CenterBounds(ir.left, ir.right, sprite_size.width - offset.x); // centered
+		int y = ir.top - sig_sprite_bottom_offset +
+				(ir.Height() + sig_sprite_size.height) / 2; // aligned to bottom
```
arbitrary line break

LordAro commented on src/road_gui.cpp:
```diff
@@ -974,15 +974,23 @@ struct BuildRoadDepotWindow : public PickerWindowBase {
+			DrawPixelInfo *old_dpi = _cur_dpi;
+			_cur_dpi = &tmp_dpi;
+			int x = (r.Width()  - ScaleSpriteTrad(64)) / 2 + ScaleSpriteTrad(31);
+			int y = (r.Height() + ScaleSpriteTrad(48)) / 2 - ScaleSpriteTrad(31);
```
48? everything else uses 58

LordAro commented on src/roadveh_cmd.cpp:
```diff
@@ -94,10 +94,10 @@ int RoadVehicle::GetDisplayImageWidth(Point *offset) const
 	}
-	return ScaleGUITrad(this->gcache.cached_veh_length * reference_width / VEHICLE_LENGTH);
+	return ScaleSpriteTrad(this->gcache.cached_veh_length * reference_width / VEHICLE_LENGTH);
 }
```
these functions seem gui-related, perhaps move to roadveh_gui.cpp ?

LordAro commented on src/settings_gui.cpp:
```diff
@@ -475,10 +465,45 @@ struct GameOptionsWindow : Window {
+				break;
+
+			case WID_GO_GUI_SCALE_AUTO:
+			{
```
unnecessary brace? (and on the wrong line) (also unnecessary braces for WID_GO_GUI_SCALE_BEVEL_BUTTON)

LordAro commented on src/settings_gui.cpp:
```diff
@@ -57,22 +57,6 @@ static const StringID _autosave_dropdown[] = {
-	INVALID_STRING_ID,
-};
-
 static Dimension _circle_size; ///< Dimension of the circle +/- icon. This is here as not all users are within the class of the settings window.
```
does this still need to be global?

LordAro commented on src/station_gui.cpp:
```diff
@@ -809,7 +811,7 @@ static const NWidgetPart _nested_station_view_widgets[] = {
 static void DrawCargoIcons(CargoID i, uint waiting, int left, int right, int y)
 {
-	int width = ScaleGUITrad(10);
+	int width = ScaleSpriteTrad(10);
```
10 !

LordAro commented on src/textfile_gui.cpp:
```diff
@@ -148,17 +148,15 @@ void TextfileWindow::SetupScrollbars(bool force_reflow)
 
 	DrawPixelInfo new_dpi;
-	if (!FillDrawPixelInfo(&new_dpi, x, y, right - x + 1, bottom - y + 1)) return;
+	if (!FillDrawPixelInfo(&new_dpi, fr.left, fr.top, fr.Width(), fr.Height())) return;
```
FillDrawPixelInfo should take a Rect param?

LordAro commented on src/town_gui.cpp:
```diff
@@ -196,28 +205,33 @@ struct TownAuthorityWindow : Window {
+				break;
+
+			case WID_TA_COMMAND_LIST:
+			{
```
bad linebreak

LordAro commented on src/town_gui.cpp:
```diff
@@ -420,26 +434,30 @@ struct TownViewWindow : Window {
 				SetDParam(3, this->town->goal[i]);
 			}
-			DrawString(cargo_text_left, cargo_text_right, y += FONT_HEIGHT_NORMAL, string);
+			DrawString(tr.Indent(20, rtl), string);
```
20?

LordAro commented on src/tree_gui.cpp:
```diff
@@ -58,8 +58,8 @@ static Dimension GetMaxTreeSpriteSize()
 	/* Avoid to use it uninitialized */
-	size.width = 32; // default width - WD_FRAMERECT_LEFT
-	size.height = 39; // default height - BUTTON_BOTTOM_OFFSET
+	size.width = ScaleGUITrad(32); // default width - WD_FRAMERECT_LEFT
+	size.height = ScaleGUITrad(39); // default height - BUTTON_BOTTOM_OFFSET
```
these constants don't exist anymore

LordAro commented on src/vehicle_gui.cpp:
```diff
@@ -3355,15 +3344,15 @@ void SetMouseCursorVehicle(const Vehicle *v, EngineImageType image_type)
 			GetCustomRotorSprite(Aircraft::From(v), image_type, &seq);
 			if (!seq.IsValid()) seq.Set(SPR_ROTOR_STOPPED);
-			y_offset = - ScaleGUITrad(5);
+			y_offset = - ScaleSpriteTrad(5);
```
space between `-` and number (function call) is a bit odd (i know, it was already there)

LordAro commented on src/widget.cpp:
```diff
@@ -24,6 +24,208 @@
+	return ScaleGUITrad(d);
+}
+
+void SetupWidgetDimensions()
```
doccomments? (+ several others)

LordAro commented on src/widget.cpp:
```diff
@@ -393,11 +605,17 @@ static inline void DrawVerticalScrollbar(const Rect &r, Colours colour, bool up_
 
+	/* track positions */
+	int left  = r.left + r.Width() * 3 / 11;
+	int right = r.left + r.Width() * 8 / 11;
```
arbitrary numbers? (that are repeated in DrawHorizontalScrollbar...)

LordAro commented on src/widget_type.h:
```diff
@@ -920,8 +923,7 @@ struct NWidgetPartWidget {
  */
-struct NWidgetPartPaddings {
-	uint8 top, right, bottom, left; ///< Paddings for all directions.
+struct NWidgetPartPaddings : RectPadding {
 };
```
Any particular reason for leaving it as a distinct struct, rather than making it a typedef?

LordAro commented on src/zoom_func.h:
```diff
@@ -99,34 +99,24 @@ static inline ZoomLevel UnScaleZoomGUI(ZoomLevel value)
 
-/**
- * Scale traditional pixel dimensions to GUI zoom level.
+/*
```
doxyment removal?

LordAro reviewed (commented):
Just idly scrolled through the whole diff. These are the thoughts I had along the way. Feel free to ignore any/all of it

LordAro reviewed (approved):
#Shipit
